### PR TITLE
Vizier output transforms

### PR DESCRIPTION
**Pull request by CompRhys**
## Motivation

I wanted to experiment with the vizier output transforms in botorch - in particular the infeasible transform. Potentially misplaced effort as it appears that within the meta ecosystem Ax actually handles this sort of stuff and so these may not actually end up being able to used very much but worth upstreaming.

## Test Plan

Added tests for all functionality, not sure I am at 100% coverage

## Related PRs

this branch takes off from https://github.com/pytorch/botorch/pull/2636 so that would need to be merged first.

### CODE REVIEW

File: botorch/models/transforms/outcome.py [hunk 1]
```diff
@@ -823,3 +828,523 @@ def untransform_posterior(self, posterior: Posterior) -> TransformedPosterior:
...
+
+                    # save intermediate values for untransform
+                    self._original_label_medians[*batch_idx, dim] = y_median
+                    self._unique_labels[(*batch_idx, dim)] = unique_y
+                    self._warped_labels[(*batch_idx, dim)] = (rank_ppf + y_median)[
+                        is_finite_mask
+                    ][unique_indices]
+
+            self._is_trained = torch.tensor(True)
+
```
Comment (saitcakmak) [inline botorch/models/transforms/outcome.py:1210]: What happens if `self.training=False`?

File: botorch/models/transforms/outcome.py [hunk 2]
```diff
@@ -823,3 +828,523 @@ def untransform_posterior(self, posterior: Posterior) -> TransformedPosterior:
...
+                ]
+            )
+            for batch_idx in batch_indices:
+                y = Y_utf[*batch_idx, :, dim].clone()
+                unique_labels = self._unique_labels[(*batch_idx, dim)]
+                warped_labels = self._warped_labels[(*batch_idx, dim)]
+
+                # Process values below median
+                below_median = y < self._original_label_medians[*batch_idx, dim]
+                if below_median.any():
```
Comment (saitcakmak) [inline botorch/models/transforms/outcome.py:1253]: This is quite a complicated process to untransform. I find it quite difficult to review since I don't know what the algorithm is supposed to do.

**Review by saitcakmak — CHANGES_REQUESTED** [review]
Thanks for adding these. I've been curious to try out the Vizier transforms. These transforms try to ensure we model the good regions better, which seems better than trying to model everything equally well, when the goal is to optimize a function. Have you tried these on any benchmarks yet? It'd be good to see how they work e2e.

None of these transforms implement the `untransform_posterior` method, which is necessary to make sure the transforms work as an `outcome_transform` in BoTorch models. Without it, `Model.posterior` will error out. Would it be possible to implement `untransform_posterior`, likely returning a `TransformedPosterior` object.

None of the transforms seem to support `Yvar` either. We use BoTorch through Ax and try to keep the same methods regardless of whether the noise is observed. What does Vizier do with observation noise? Is it simply ignored, or is this just a limitation of current implementation?

**Comment by CompRhys** [comment]
> Thanks for adding these. I've been curious to try out the Vizier transforms. These transforms try to ensure we model the good regions better, which seems better than trying to model everything equally well, when the goal is to optimize a function. Have you tried these on any benchmarks yet? It'd be good to see how they work e2e.

I implemented these here before looking into what I would need to do to connect them to my benchmark testcase that I have been prototyping in Ax. I assumed that the best case would be to implement it lower in the stack and then bubble up but seemingly I should have implemented at the level of Ax (especially as the harder part here was handling the `batch_size`) hence the comment about potentially misplaced effort. It may be that these shouldn't be merged here if the details can't be hashed out.

> None of these transforms implement the `untransform_posterior` method, which is necessary to make sure the transforms work as an `outcome_transform` in BoTorch models. Without it, `Model.posterior` will error out. Would it be possible to implement `untransform_posterior`, likely returning a `TransformedPosterior` object.

This should be able to be done based on the logic in `Standardize` and just using the sample transform. In general the bit here that makes things annoying is the `batch_size`. 

> None of the transforms seem to support `Yvar` either. We use BoTorch through Ax and try to keep the same methods regardless of whether the noise is observed. What does Vizier do with observation noise? Is it simply ignored, or is this just a limitation of current implementation?

I think the vizier approach is similar to performing the transformations at the layer of Ax. Their equivalent to Metric only deals with a single float observation. As such and from the docs I believe it's ignored. I don't think that we can implement Yvar for either the LogWarping or the HalfRank transform

File: botorch/models/transforms/outcome.py (hunk 2 above)
Comment (CompRhys) [reply botorch/models/transforms/outcome.py:1253]: The HalfRankTransform splits the data at the median and then warps the bottom half onto a gaussian whose std matches the std of the top half of the distribution. On inverting there are 3 cases, a) we can just look it up as it's very close to a known value from the fit, b) we take a in support linear interpolation, c) outside of the support we take a linear extrapolation based on the entire data range.

File: botorch/models/transforms/outcome.py [hunk 3]
```diff
@@ -823,3 +828,523 @@ def untransform_posterior(self, posterior: Posterior) -> TransformedPosterior:
...
+        return tensor.nan_to_num(max_value).min()
+    return tensor.nan_to_num(max_value).min(dim=dim, keepdim=keepdim)
+
+
+def _check_batched_output(Y: Tensor, batch_shape: Tensor) -> None:
+    """Utility for common output transform checks."""
+    if Y.shape[:-2] != batch_shape:
+        raise RuntimeError(
+            f"Expected Y.shape[:-2] to be {batch_shape}, matching "
+            "the `batch_shape` argument to `Standardize`, but got "
```
[the remark this answers is not in the document]
Comment (CompRhys) [reply botorch/models/transforms/outcome.py:856]: copy-paste issue

File: botorch/models/transforms/outcome.py [hunk 4]
```diff
@@ -823,3 +828,523 @@ def untransform_posterior(self, posterior: Posterior) -> TransformedPosterior:
...
+
+            self._is_trained = torch.tensor(True)
+
+        # Expand warped_bad_value to match Y's shape
+        expanded_bad_value = self.warped_bad_value.unsqueeze(-2).expand(
+            *Y.shape[:-2], Y.shape[-2], -1
+        )
+        expanded_shift = self._shift.unsqueeze(-2).expand(
+            *Y.shape[:-2], Y.shape[-2], -1
+        )
```
[the remark this answers is not in the document]
Comment (CompRhys) [reply botorch/models/transforms/outcome.py:929]: `expand_as` is new to me but is cleaner so will swap to that.

File: botorch/models/transforms/outcome.py [hunk 5]
```diff
@@ -823,3 +828,523 @@ def untransform_posterior(self, posterior: Posterior) -> TransformedPosterior:
...
+    Inspired by output-space transformations in Vizier: https://arxiv.org/abs/2408.11527
+    """
+
+    def __init__(
+        self, batch_shape: torch.Size | None = None, offset: float = 1.5
+    ) -> None:
+        """Initialize transform.
+
+        Args:
+            offset: Offset parameter for the log transformation. Must be > 0.
```
[the remark this answers is not in the document]
Comment (CompRhys) [reply botorch/models/transforms/outcome.py:984]: The larger the value of the offset `s` the more of the resulting range will be used in differentiating good values.

File: botorch/models/transforms/outcome.py [hunk 6]
```diff
@@ -823,3 +828,523 @@ def untransform_posterior(self, posterior: Posterior) -> TransformedPosterior:
...
+            A two-tuple with the transformed outcomes:
+            - The transformed outcome observations.
+            - The transformed observation noise (if applicable).
+        """
+        _check_batched_output(Y, self._batch_shape)
+
+        if Yvar is not None:
+            raise NotImplementedError(
+                "LogWarperTransform does not support transforming observation noise"
+            )
```
[the remark this answers is not in the document]
Comment (CompRhys) [reply botorch/models/transforms/outcome.py:1015]: Yes, in general I think this basically always the case with warping and other manipulations. 

If you start thinking about handling these things by sampling and then propagating individual samples then I think that in the general case the warping applied to Y with Y playing the role of the mean is not going to be the same as the mean of the warped samples.

File: botorch/models/transforms/outcome.py [hunk 7]
```diff
@@ -823,3 +828,523 @@ def untransform_posterior(self, posterior: Posterior) -> TransformedPosterior:
...
+
+        Returns:
+            A two-tuple with the transformed outcomes:
+            - The transformed outcome observations.
+            - The transformed observation noise (if applicable).
+        """
+        if Yvar is not None:
+            raise NotImplementedError(
+                "HalfRankTransform does not support transforming observation noise"
+            )
```
[the remark this answers is not in the document]
Comment (CompRhys) [reply botorch/models/transforms/outcome.py:1135]: I can't see any reference to it in the docs. I don't think they handle it.

File: botorch/models/transforms/outcome.py [hunk 8]
```diff
@@ -823,3 +828,523 @@ def untransform_posterior(self, posterior: Posterior) -> TransformedPosterior:
...
+                    else [  # this allows it to work with no batch dim
+                        (...,),
+                    ]
+                )
+                for batch_idx in batch_indices:
+                    y_median = Y_medians[*batch_idx, dim]
+                    y = Y_transformed[*batch_idx, :, dim]
+
+                    # Get finite values and their ranks for each batch
+                    is_finite_mask = ~torch.isnan(y)
```
[the remark this answers is not in the document]
Comment (CompRhys) [reply botorch/models/transforms/outcome.py:1165]: I think this is the salient point, for me which indicates that I should probably have done this in Ax. Would Ax error out if a metric returned NaN before applying transforms or only after?

File: botorch/models/transforms/outcome.py [hunk 9]
```diff
@@ -823,3 +828,523 @@ def untransform_posterior(self, posterior: Posterior) -> TransformedPosterior:
...
+                    # TODO: this is annoying but torch.unique doesn't support
+                    # returning indices
+                    np_unique_y, np_unique_indices = np.unique(
+                        y[is_finite_mask].numpy(), return_index=True
+                    )
+                    unique_y = torch.from_numpy(np_unique_y)
+                    unique_indices = torch.from_numpy(np_unique_indices)
+
+                    for i, val in enumerate(unique_y):
+                        ranks[y == val] = i + 1
```
[the remark this answers is not in the document]
Comment (CompRhys) [reply botorch/models/transforms/outcome.py:1177]: I think it's worth having this generality incase other packages like `BoFire` want to make use of these methods, I tried to stick as close to the reference implementation as possible but expanding to handle multi-dimension and `batch_size`

File: botorch/models/transforms/outcome.py [hunk 10]
```diff
@@ -823,3 +828,523 @@ def untransform_posterior(self, posterior: Posterior) -> TransformedPosterior:
...
+
+            # Estimate the relative frequency of feasible points
+            p_feasible = (0.5 + num_feasible) / (1 + Y.numel())
+
+            self.warped_bad_value = warped_bad_value
+            self._shift = -torch.nanmean(Y, dim=-2) * p_feasible - warped_bad_value * (
+                1 - p_feasible
+            )
+
+            self._is_trained = torch.tensor(True)
```
[the remark this answers is not in the document]
Comment (saitcakmak) [reply botorch/models/transforms/outcome.py:921]: Actually, looks like we need to register it as a buffer if we want it in the state dict

File: botorch/models/transforms/outcome.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (saitcakmak) [reply botorch/models/transforms/outcome.py:1015]: Yeah, definitely. Transforming the noise is a general problem with non-linear transforms. We try to approximate it somehow (with moment matching etc) to have support for noise observations.

File: botorch/models/transforms/outcome.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (saitcakmak) [reply botorch/models/transforms/outcome.py:1135]: Interesting. We find it useful to specify the observed noise (if available) since this can lead to better model fits.

**Comment by saitcakmak** [comment]
> Ax vs BoTorch

This is something I've also been thinking about quite a bit lately, as part of reworking what Ax uses by default (though this has been rather focused on input transforms so far). Outcome transforms may be slightly simpler in Ax. You technically only need to be able to transform, though we also want some rough untransform method for reporting & cross validation purposes. In BoTorch, you need the transform to be differentiable, since the untransformed posterior is used to compute the acquisition value which needs to be differentiated through during optimization. Ax transforms are applied once as a pre-processing step but BoTorch transforms are applied at each `posterior` call. There is also the aspect of learnable / tunable transforms in BoTorch but that doesn't apply to outcome transforms.

If I were to implement Vizier transforms, I'd probably try them in Ax, since there are fewer restrictions.

> In general the bit here that makes things annoying is the batch_size.

Do you often use batched models? If not, you don't need to support `batch_size` arg and different treatment for each batch. You could use the same parameters for all batches.

> I think the vizier approach is similar to performing the transformations at the layer of Ax. 

Agreed. It is more of a pre-processing step.

> Their equivalent to Metric only deals with a single float observation. As such and from the docs I believe it's ignored. I don't think that we can implement Yvar for either the LogWarping or the HalfRank transform

Cool, this is good to know.

**Comment by CompRhys** [comment]
Will move forward with a PR at Ax as that seems like the faster way to make progress on my usecase

**Comment by saitcakmak** [comment]
Sounds good. [StandardizeY](https://github.com/facebook/Ax/blob/main/ax/modelbridge/transforms/standardize_y.py) or [PowerTransformY](https://github.com/facebook/Ax/blob/main/ax/modelbridge/transforms/power_transform_y.py) should be good examples to follow

**Comment by CompRhys** [comment]
closing as abandoned
